src/Build/Construction/Solution/SolutionProjectGenerator.cs

```diff
@@ -782,7 +782,7 @@ private void EvaluateAndAddProjects(List<ProjectInSolution> projectsInOrder, Lis
                 AddTraversalTargetForProject(traversalInstance, project, projectConfiguration, "Publish", null, canBuildDirectly);
 
                 // Add any other targets specified by the user that were not already added
-                foreach (string targetName in _targetNames.Except(traversalInstance.Targets.Keys, StringComparer.OrdinalIgnoreCase))
+                foreach (string targetName in _targetNames.Where(i => !traversalInstance.Targets.ContainsKey(i)))
```

Comment:
Should we extract method that returns the enumerable? And use it in four places? And add a detailed comment explaining that it has to be lazy because evidently the traversalInstance.Targets contents changes during enumeration? It's a tricky spot.

Reply:
also need to pay attention to how Keys behaves - is it returning the live collection or a snapshot of the keys?

Reply:
Extract method might be dicey - maybe better keep as is for simplicity... just add comments? You decide

Reply:
I want this to be as small of a change and as safe of a change as possible, since this is going into 16.10 rather than 17.0. This is a straight revert of the relevant part of the commit that caused the issue, which is safest in my opinion. Adding a comment is a good plan, though—I'll just add it for the first case, since that's the only one that I _know_ should be just in time.

Reply:
Agreed on keeping this to a revert. Can extract as follow up in a future milestone if that's better.

Reply:
Yes, let’s not extract, it was just a knee jerk reaction of seeing the changes in four places. All good!